apm-agent-plugins/apm-vertx/apm-vertx-common/src/main/java/co/elastic/apm/agent/vertx/NettyByteTransfer.java

```diff
@@ -0,0 +1,60 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2021 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.vertx;
+
+import co.elastic.apm.agent.util.IOUtils;
+import io.netty.buffer.ByteBuf;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+public class NettyByteTransfer implements IOUtils.BufferTransfer {
+
+    @Nullable
+    private static NettyByteTransfer INSTANCE;
+    private final ByteBuf byteBuf;
+
+    private NettyByteTransfer(ByteBuf byteBuf) {
+        this.byteBuf = byteBuf;
+    }
+
+    public static NettyByteTransfer get(ByteBuf byteBuf) {
+        if (INSTANCE == null) {
+            INSTANCE = new NettyByteTransfer(byteBuf);
```

Comment:
Unless I'm missing something, this stores the `ByteBuf` from the first request as a singleton which prevents the capturing from other requests.
I'd suggest changing the interface so that you don't need instance variables, instead the ByteBuf would be a parameter to the `transfer` and `readableBytes` methods.

Reply:
🤦 yes, you're absolutely right. Just extracted this functionality into a Vertx-specific IOUtils that extends the common IOUtils